### PR TITLE
Reordered Musicbrainz tags as suggested in lp1924620 +  lp1899350

### PR DESCRIPTION
**Pull request by ehmic**
["Musikbrainz: Wrong column order in result window"](https://bugs.launchpad.net/mixxx/+bug/1924620)
["wierd column order in Musicbraiz result dialog"](https://bugs.launchpad.net/mixxx/+bug/1899350)

Before:

musicbrainz-before

After:

musicbrainz-after

Todo:

- Find out where the colons in 'Artist' and 'Album' stem from and get rid of them
- Optional: Modify column sizes for a better look, or, if possible, resize them after the import from Musicbrainz to make them fit to the text

Open questions:

-  Do I need to also change the column order in  `dlgtagfetcher.ui` - item `<widget class="QTreeWidget" name="submit_tree">`?
- Is `dlgtagfetcher` used for something else than Musicbrainz so that my changes may break stuff elsewhere?

### CODE REVIEW

**Comment by ronso0** [comment]
> Is dlgtagfetcher used for something else than Musicbrainz so that my changes may break stuff elsewhere?

I recommend setting up an IDE ([Eclipse tutorial on the wiki](https://github.com/mixxxdj/mixxx/wiki/Eclipse)) and inspect what calls dlgtagfetcher.
Or search the codebase for "dlgtagfetcher"

Please stick with th eprevious left-aligned columns, Title being aligned right looks weird to me.

> Find out where the colons in 'Artist' and 'Album' stem from and get rid of them

Why? This is essential track info. If the columns aren't visible users don't know what the tags are populated with.

> Optional: Modify column sizes for a better look, or, if possible, resize them after the import from Musicbrainz to make them fit to the text

maybe this helps
https://doc.qt.io/qt-5/qheaderview.html#resizeSections
https://doc.qt.io/qt-5/qheaderview.html#ResizeMode-enum

**Comment by ronso0** [comment]
> I recommend setting up an IDE (Eclipse tutorial on the wiki) and inspect what calls dlgtagfetcher.

Worth the effort if you start seriously working on Mixxx : )
The initial setup may be a bit of a hazzle, though I consider it worth it because Eclipse is quite powerful.
There are other IDEs, though I don't know which are available for your OS.

**Comment by uklotzde** [comment]
@ronso0 "colon", not "column" ;)

File: src/library/dlgtagfetcher.cpp [hunk 1]
```diff
@@ -86,12 +88,12 @@ void DlgTagFetcher::init() {
...
-    results->setColumnWidth(2, 160); // Album artist column
-    results->setColumnWidth(3, 50);  // Track (numbers) column
-    results->setColumnWidth(4, 160); // Title column
-    results->setColumnWidth(5, 160); // Artist column
+    results->setColumnWidth(0, 160); // Title column
+    results->setColumnWidth(1, 160); // Artist column
+    results->setColumnWidth(2, 160); // Album column
+    results->setColumnWidth(3, 50);  // Year column
+    results->setColumnWidth(4, 50);  // Track (numbers) column
+    results->setColumnWidth(5, 160); // Album artist column
```
Comment (ronso0) [inline src/library/dlgtagfetcher.cpp:96]: that is just the initial view.
Adjsuting the columns later on might work with
`results->header()->resizeSections(QHeaderView::ResizeToContents);` somewhere in updateStack() I suppose.

**Comment by ehmic** [comment]
> > Is dlgtagfetcher used for something else than Musicbrainz so that my changes may break stuff elsewhere?
> 
> I recommend setting up an IDE ([Eclipse tutorial on the wiki](https://github.com/mixxxdj/mixxx/wiki/Eclipse)) and inspect what calls dlgtagfetcher.
> Or search the codebase for "dlgtagfetcher"
> 

Will do...

> Please stick with the left-aligned columns, Title being aligned right looks weird to me.
> 
Strangely enough I did not touch the column alignment. But I agree, this doesn't look good - will change it accordingly .

> > Find out where the colons in 'Artist' and 'Album' stem from and get rid of them
> 
> Why? This is essential track info. If the columns aren't visible users don't know what the tags are populated with.
> 

Not sure if we talk about the same here.
I'm mean those colons / double dots in the column header... only 'Artist' and 'Album' have these. In the file 'dlgtagfetcher.ui' they are setup correctly, but somehow they get altered somewhere afterwards.

> > Optional: Modify column sizes for a better look, or, if possible, resize them after the import from Musicbrainz to make them fit to the text
> 
> maybe this helps
> https://doc.qt.io/qt-5/qheaderview.html#resizeSections
> https://doc.qt.io/qt-5/qheaderview.html#ResizeMode-enum

Thanks, will see what I can do!

**Comment by ronso0** [comment]
> Strangely enough I did not touch the column alignment. But I agree, this doesn't look good - will change it accordingly

that can be set reliably in the ui file (unless it's touched in c++ later on)

**Comment by ehmic** [comment]
> @ronso0 "colon", not "column" ;)

Yeah, just got ninjaed   

> > I recommend setting up an IDE (Eclipse tutorial on the wiki) and inspect what calls dlgtagfetcher.
> 
> Worth the effort if you start seriously working on Mixxx : )
> The initial setup may be a bit of a hazzle, though I consider it worth it because Eclipse is quite powerful.
> There are other IDEs, though I don't know which are available for your OS.

I have QtCreator installed on my machine but mainly use Geany and linux standard tools.
Maybe I'll give Eclipse a try later.

File: src/library/dlgtagfetcher.cpp (hunk 1 above)
Comment (ehmic) [reply src/library/dlgtagfetcher.cpp:96]: Thanks for this hint. I'm sure

> > Strangely enough I did not touch the column alignment. But I agree, this doesn't look good - will change it accordingly
> 
> that can be set reliably in the ui file (unless it's touched in c++ later on)

Seems ok to me in the ui file so I guess it must be altered somewhere else.

**Comment by uklotzde** [comment]
The alignment of the first column didn't change, but its content. You need to adjust the column properties in the .ui file accordingly.

**Comment by ehmic** [comment]
Oh dear, what happened here... why do I have 63 commits now? Must have done something wrong.
Is there any way to fix that?

**Comment by daschuer** [comment]
This must have happened due to a wrong merge. 
I will have a look.

**Comment by daschuer** [comment]
This should do the trick. 

git rebase --onto 1bcdbb753f840e6e24783b7bba4a732985a58ef9 d3c5b167d372b4e721cdd2dd4cf692dc545fc4ca
git push -f 

The result is two commits in this PR. Is this desired?

**Comment by ehmic** [comment]
> This should do the trick.
> 
> ```
> git rebase --onto 1bcdbb753f840e6e24783b7bba4a732985a58ef9 d3c5b167d372b4e721cdd2dd4cf692dc545fc4ca
> git push -f 
> ```
> 
> The result is two commits in this PR. Is this desired?

If this fixes the problem with this PR, yes of course - even if some of my changes get lost.
Will try that, thanks!

**Comment by ehmic** [comment]
Received an error message in the process and don't know how I should treat this problem
...
M	CMakeLists.txt
Falle zurück zum Patchen der Basis und zum 3-Wege-Merge ...
automatischer Merge von CMakeLists.txt
KONFLIKT (Inhalt): Merge-Konflikt in CMakeLists.txt
error: Merge der Änderungen fehlgeschlagen.
Anwendung des Patches fehlgeschlagen bei 0002 Update Windows build environment and remove workaround for broken debug builds
Hinweis: Benutzen Sie 'git am --show-current-patch', um den
Hinweis: fehlgeschlagenen Patch zu sehen.
Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort"


Any advice?

**Comment by daschuer** [comment]
I have just pushed my solution, so you may try
```
git reset d6fcc1d98d348950edd9090de0db4ba91d253081 --hard
git pull https://github.com/daschuer/mixxx.git musicbrainz_column_reorder
git reset 8b38a55c156b577d9effbdb10526ec5efd96b5ee --hard
git push -f
```

**Comment by ehmic** [comment]
Ok, tried that and ran into this (sorry for being such a Git noob :grimacing:) :

git push -f
fatal: Sie befinden sich im Moment auf keinem Branch.
Um die Historie, führend zum aktuellen (losgelöster HEAD)
Status zu versenden, benutzen Sie

    git push origin HEAD:<Name-des-Remote-Branches>

**Comment by Holzhaus** [comment]
`git mergetool` is your friend. Did you remove all commits that do not belong to this branch when rebasing?

**Comment by Holzhaus** [comment]
You Re not on a branch. Run `git checkout musicbrainz_column_reorder` first, then repeat the instructions from @daschuer .

**Comment by ehmic** [comment]
> You're not on a branch. Run `git checkout musicbrainz_column_reorder` first, then repeat the instructions from @daschuer .

That did the trick - thank you very much everyone!

**Comment by ehmic** [comment]
Colons in the header are still a problem - only for Artist and Album (see pictures below).
On my computer they are also present in the latest Mixxx 2.3.0-beta from the official repository. 
Can someone please confirm that they do exist?
I am more than willing to do some more investigation, but want to be sure that this is not just a problem with my setup.

With these column-settings in the ui file....

musicbrainz-original-ui

I get this ....

musicbrainz-original

When I change it to something else...

musicbrainz-test-ui

... the colons are gone...

musicbrainz-test

And as soon as I change the columns back the colons re-appear

**Comment by ehmic** [comment]
Found the root cause which make the colons appear.  
The translations for Album and Artist contain those colons in file `mixxx_en_GB.ts`
I changed the file accordingly and ran` lrelease mixxx_en_GB.ts `afterwards to re-generate the `mixxx_en_GB.qm` file.
Not sure if this was the correct way to do it, but at least that fixed the issue for me.

Now, before I proceed, I have few questions:

- If this was the correct way, should I commit both, the `mixxx_en_GB.ts` AND the `mixxx_en_GB.qm` file, or is the `ts` file sufficient because the `.qm` file will be automatically re-generated anyhow? 
- Is it generally OK if I use this Musicbrainz PR here to also fix translations?

Asking the latter question also because I found another translation issue which is not related to MusicBrainz, but recording preferences. Should I create a new PR for this one or may I, as an exception, also fix this problem in this PR here?

Recording preferences (album holds colon):

record_translate

**Comment by ronso0** [comment]
> Should I create a new PR for this one or may I, as an exception, also fix this problem in this PR here?

yes, that is a separate issue.

**Comment by ronso0** [comment]
re: colons in en_GB
I just removed those three occurences manually on transifex. should be fixed next time the translations are pulled into the mixxx repo.

**Comment by ehmic** [comment]
Thanks for removing the colons in transifex. 

This is the latest version ...

musicbrainz-original

**Review by daschuer — APPROVED** [review]
This seems to be a low risk PR suitable for being part in 2.3.0
LGTM 

I have pulled the : fix from transiflex already.
@ronso0 what do you think?

**Comment by Holzhaus** [comment]
Shouldnt track number be the first, and Album artist between artist and album? But I donvt want to bikes he'd about this.

**Comment by ehmic** [comment]
> Shouldnt track number be the first, and Album artist between artist and album? But I donvt want to bikes he'd about this.

For me usually the Title, Artist and Album is the most important info needed to identify a track - the track number may be relevant too in some cases.
When I reorder everything in the way you suggested, the main Information which helps to identify a track (Album in this case) may be invisible until the bottom scroll-bar is used. 
So with keeping efficiency in mind, maybe it is better to, if anything, only re-order track number?

As an addition - I noticed that quite a lot of the purchased tracks on my computer have empty Artist Album fields anyhow.

**Comment by ronso0** [comment]
yes, looks like low risk. I hope I can take a closer look tonight when I'm back at my pc

**Comment by ronso0** [comment]
LGTM.
Unfortunately `results->header()->resizeSections(QHeaderView::ResizeToContents);` which I suggested may not be appreciated I'm afraid: while it does resize the the table attempting to make everything fit it does
* produce a rather stuffed view, the inter-column gap is hard to recognize (OS theme-dependent?)
* adjusted column widths are reset when proceeding to the next/previous track

Maybe we should revert that so the custom column layout is locked per session like before?
Maybe issues is fixed with a few pixel padding-right (for each column)?

Side note not holding up this PR: are all tags shown which are provided by MusicBrainz (and would override the user set)?

**Comment by ehmic** [comment]
> LGTM.
> Unfortunately `results->header()->resizeSections(QHeaderView::ResizeToContents);` which I suggested may not be appreciated I'm afraid: while it does resize the the table attempting to make everything fit it does
> 
>     * produce a rather stuffed view, the inter-column gap is hard to recognize (OS theme-dependent?)
> 
>     * adjusted column widths are reset when proceeding to the next/previous track

Agree with you on the first one, it looks indeed a little bit stuffed.  But isn't the latter the expected behaviour? 
Shouldn't the column widths 'adapt' to the new content size if next/previous track is selected?

I noticed that setting `headerStretchLastSection=false` can also lead to unwanted results.

> 
> 
> Maybe we should revert that so the custom column layout is locked per session like before?
> Maybe issues is fixed with a few pixel padding-right (for each column)?

Just investigated the problem and maybe have another solution.

How about this:

- Set `headerStretchLastSection` back to `true`

- Resize each row to their contents, plus a margin

```
 for (int i = 0; i <= results->model()->rowCount(); i++) {
        results->resizeColumnToContents(i);
        int sectionSize = (results->columnWidth(i) + 10); // margin of 10
        results->header()->resizeSection(i,sectionSize);
    }
```
Here is the result:

mbsizemargin

... headerstretchLastSection also does look better ...

mbstretchls


> 
> Side note not holding up this PR: are all tags shown which are provided by MusicBrainz (and would override the user set)?

Not sure about this one

**Comment by ronso0** [comment]
>  for (int i = 0; i <= results->model()->rowCount(); i++) {
        results->resizeColumnToContents(i);
        int sectionSize = (results->columnWidth(i) + 10); // margin of 10
        results->header()->resizeSection(i,sectionSize);
    }

Yes, that works fine. Just implement it exactly as suggested, then this PR is ready IMO.

**Comment by ehmic** [comment]
Thanks for the reviews!
Will implement as mentioned above, but with an exception - `(int i = 0; i <= results->model()->rowCount(); i++)` needs of course to be actually `(int i = 0; i < results->model()->columnCount(); i++)` (silly me, sorry for confusing this) 

I found one issue which I was not able to solve by now - but am still trying.
It is a minor one, so maybe it can be neglected if no-one has a solution?

Here is the problem:
With my solution, if the size of the largest content in a section deceeds a certain limit, the resulting resized section is slightly bigger than it could/should be. This is mostly the case for the 'Track' section, but can also happen in other ones, like 'Title'.

Two examples:

Here only the 'Track' section deceeds the limit....

test3

... and here it is the 'Title' section as well...

test2


Have a final 'cleanup' question

I don't see any case where the following code is important anymore. 
Am I right that there is no need for it and it can be safely removed?

   `// Resize columns, this can't be set in the ui file`
   `results->setColumnWidth(0, 160); // Title column`
   `results->setColumnWidth(1, 160); // Artist column`
   `results->setColumnWidth(2, 160); // Album column`
   `results->setColumnWidth(3, 50);  // Year column`
   `results->setColumnWidth(4, 5);  // Track (numbers) column`
   `results->setColumnWidth(5, 160); // Album artist column`

Thanks again and sorry for creating so much noise for such a 'minor' PR!

**Comment by ronso0** [comment]
> With my solution, if the size of the largest content in a section deceeds a certain limit, the resulting resized section is slightly bigger than it could/should be. This is mostly the case for the 'Track' section, but can also happen in other ones, like 'Title'.

I think we can ignore that as long as it works fine for other columns.

Yes, I think default column widths can be removed.
If the colum value (content) is empty the column title would be used to calculate the minimal width, right?

**Comment by daschuer** [comment]
Since this works good, I like to merge this and proceed to the release. 
Thank you very much.

**Comment by daschuer** [comment]
> Yes, I think default column widths can be removed.
> If the colum value (content) is empty the column title would be used to calculate the minimal width, right?

Yes that is the case. I have removed it in 65ae0bf

**Pull request merged**